### PR TITLE
#448 CSV import in matrix overview

### PR DESCRIPTION
**Pull request by hehu80**

### CODE REVIEW

File: src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:1]: Evaluate upsert, maybe deletion is too rigorous

**Review by KlausRicharz — COMMENTED** [review]
Maybe better store timereports in a single call to timereport service

File: src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java
Comment (KlausRicharz) [reply src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:1]: what about other bookings that are not part of your project. so maybe better to delete only those that are matching the orders in the file?

Comment (hehu80) [reply src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:1]: Or we let the user decide if he wants to keep existing bookings or not: 
Screenshot 2024-11-17 at 11-35-01 SALAT - Matrixübersicht

File: src/main/java/org/tb/dailyreport/action/ShowMatrixAction.java [hunk 1]
```diff
@@ -101,6 +114,15 @@ public ActionForward executeAuthenticated(ActionMapping mapping, ShowMatrixForm
             }
         }
 
+        if(task == null && reportForm.getImportFile() != null) {
```
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/action/ShowMatrixAction.java:117]: Should we introduce a specific task "importCsv"?

File: src/main/java/org/tb/dailyreport/rest/DailyReportData.java [hunk 2]
```diff
@@ -56,4 +56,8 @@ static DailyReportData valueOf(TimereportDTO timeReport) {
                 .orderSign(timeReport.getCustomerorderSign())
                 .build();
     }
+
+    public DailyReportData withoutId(){
+        return toBuilder().id(0).build();
```
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/rest/DailyReportData.java:61]: 0 = special case?

File: src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    doCreateDailyReports(report.getDate(), employeeOrderId, bookingsOfOrder, upsert);
+                });
+    }
+
+    private void doCreateWorkingDay(DailyWorkingReportData report, Employeecontract employeecontract) {
+        var existingWorkingDay = ofNullable(workingdayDAO.getWorkingdayByDateAndEmployeeContractId(
+                report.getDate(), requireNonNull(employeecontract.getId(), "ID of contract is required")));
+
+        var workingDay = existingWorkingDay.orElseGet(Workingday::new);
+        workingDay.setEmployeecontract(employeecontract);
```
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:76]: Das ist nur bei neuen Working Days wichtig

File: src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            workingDay.setStarttimeminute(0);
+        });
+        workingDay.setType(report.getType());
+        workingdayService.upsertWorkingday(workingDay);
+    }
+
+    private void doCreateDailyReports(LocalDate day, Long employeeOrderId, List<DailyReportData> bookings, boolean upsert) {
+        var employeeOrder = employeeorderDAO.getEmployeeorderById(employeeOrderId);
+        if (employeeOrder == null) {
+            throw new ResponseStatusException(NOT_FOUND, "Could not find employeeorder with id " + employeeOrderId);
```
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:99]: ResponseStatusException nicht im Service verwenden, besser InvalidDataException.

File: src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        var employeeContract = employeeOrder.getEmployeecontract();
+        if (employeeContract == null) {
+            throw new ResponseStatusException(NOT_FOUND, "Could not find employeecontract for order with id " + employeeOrderId);
```
Comment (KlausRicharz) [inline src/main/java/org/tb/dailyreport/service/DailyWorkingReportService.java:104]: same -> InvalidDataException

**Review by KlausRicharz — CHANGES_REQUESTED** [review]
see comments

**Review by KlausRicharz — APPROVED** [review]

**Pull request merged**